inference quickstart (#19)
E3SM-Project/aigroup: docs/ace2-inference.md

```console
git clone https://huggingface.co/allenai/ACE2-EAMv3
```

!!! note "git lfs"
    If you run into issues related to git lfs, you may need to install that.

### 3. Run Inference

Navigate to the `ace` repository directory:

```console
cd ace
```

Create a configuration file named `config-inference.yaml` with the following content. Make sure to update the paths to match your environment.

``` { .yaml .annotate }
experiment_dir: /pscratch/sd/m/mahf708/ACE2-EAMv3/test1 # (1)!
n_forward_steps: 1458 # (2)!
forward_steps_in_memory: 80 # (3)!
checkpoint_path: /pscratch/sd/m/mahf708/ACE2-EAMv3/ace2_EAMv3_ckpt.tar # (4)!
initial_condition: # (5)!
  path: /pscratch/sd/m/mahf708/ACE2-EAMv3/initial_conditions/1971010100.nc
  start_indices:
    n_initial_conditions: 2
    first: 0
    interval: 1
forcing_loader: # (6)!
  dataset:
    data_path: /pscratch/sd/m/mahf708/ACE2-EAMv3/forcing_data/
  num_data_workers: 2
logging: # (7)!
  log_to_screen: true
  log_to_wandb: false
  log_to_file: true
data_writer: # (8)!
  save_prediction_files: true
```

1. **Output directory** — All inference outputs (predictions, diagnostics, logs) are saved here. Create this directory before running.

> 2. **Number of forward steps** — Total timesteps to run. Each step is 6 hours, so 1458 steps ≈ 365 days. Note that there is a condition internal to ACE necessitating that this number is two steps short of the side of the full data set (in this example, a year; see forcing_loader)

Author: ```suggestion
2. **Number of forward steps** — Total timesteps to run. Each step is 6 hours, so 1458 steps ≈ 365 days. See note below on limitation.
```

3. **Steps in memory** — Batch size for GPU memory. Lower this if you run into OOM errors. 80 is a good default.
4. **Model checkpoint** — Path to the pretrained ACE2-EAMv3 weights (`.tar` file from Hugging Face).
5. **Initial conditions** — Starting atmospheric state. `n_initial_conditions` runs multiple ensemble members; `first` and `interval` control which samples to use from the IC file.
6. **Forcing data** — External forcing (SST, solar, GHGs, etc.). The loader reads Zarr/NetCDF files from this path. `num_data_workers` controls parallel I/O.
7. **Logging options** — `log_to_screen` prints progress; `log_to_wandb` sends metrics to Weights & Biases (requires login); `log_to_file` saves to `inference_out.log`.

> 8. **Output writer** — Set `save_prediction_files: true` to write NetCDF outputs. Set to `false` for validation-only runs.

Author: ```suggestion
8. **Output writer** — Set `save_prediction_files: true` to write NetCDF outputs. Set to `false` for validation-only runs. Additionally, one could `names: [T_4, T_5]` to request only `T_4` and `T_5` in the output.
```

    ```

!!! tip "uv cache"
    Sometimes, you will need to the enviornment variable `UV_CACHE_DIR`, e.g., on NERSC, `export UV_CACHE_DIR="$PSCRATCH/.cache/uv"`

### 4. Results

The results will be saved in the `experiment_dir` specified in the config file. The output directory structure will look like this:

```console
> ls /pscratch/sd/m/mahf708/ACE2-EAMv3/test1 -1 
annual_diagnostics.nc
autoregressive_predictions.nc
autoregressive_target.nc
config.yaml
inference_out.log
initial_condition.nc
mean_diagnostics.nc
monthly_mean_predictions.nc
monthly_mean_target.nc
restart.nc
time_mean_diagnostics.nc  
```

where the autoregressive_predictions.nc file has the following header:

```console
ncdump -h /pscratch/sd/m/mahf708/ACE2-EAMv3/test1/autoregressive_predictions.nc
netcdf autoregressive_predictions {
dimensions:
        time = UNLIMITED ; // (1458 currently)
        sample = 2 ;
        lat = 180 ;
        lon = 360 ;
variables:
        int64 time(time) ;
                time:units = "microseconds" ;
        int64 init_time(sample) ;
                init_time:units = "microseconds since 1970-01-01 00:00:00" ;
                init_time:calendar = "noleap" ;
        ...
```

> These are the available variables:

Reviewer: You can also specify what variables and time frequency to save. I think this is probably a highly used config, maybe we should have an example?

Author: I didn't know about this feature. Let me investigate it and document it here
